stream: use `Math.floor()` instead of `parseInt()`

addaleax opened the pull request:
Do not use `parseInt()`, since the argument is typically either
`undefined` or an integer, but `parseInt()` expects a string
argument. Also, be explicit about the fact that a value of `NaN`
is a common scenario.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mscdex commented:
Be aware there is a change in behavior with this when it comes to converting strings to a numeric value. `Math.floor()` interprets the various supported numerical notations and can end up interpreting the string differently depending on those notations (e.g. `'0xfff'`, `'2e10'`), whereas `parseInt(str, 10)` will stop parsing the string when it finds a non-digit character (not including a possible `'-'` or `'+'` prefix) and will interpret those first `n` consecutive digits as a base-10 integer.

There are also differences when passing non-string values too. For example: `Math.floor(Infinity)` will return `Infinity` while `parseInt(Infinity, 10)` will return `NaN`. Another example: `Math.floor(null)` will return `0` while `parseInt(null, 10)` will return `NaN`.

addaleax commented:
I’ve rebased this.

@mscdex Yeah, I know that this changes behaviour for some input values. I’ve replaced the dont-land-on tags with semver-major, assuming that that is what you feel more comfortable with here (@nodejs/tsc for reviews).

mcollina commented:
If there is a change of behavior, I’d like to have it covered by a test.

addaleax commented:
@mcollina I’ve added a test, PTAL.

addaleax commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1804/

mscdex commented on test/parallel/test-stream-readable-read-string.js:
```diff
@@ -0,0 +1,13 @@
+
+const readable = new Readable({ read() {} });
+readable.push('a'.repeat(100));
+assert.strictEqual(readable.read('1e2').length, 100);
```
Might test non-string values like `null` and `Infinity` as well.

BridgeAR commented on lib/_stream_readable.js:
```diff
@@ -381,7 +382,7 @@ function howMuchToRead(n, state) {
 // You can override either this method, or the async _read(n) below.
 Readable.prototype.read = function(n) {
   debug('read', n);
-  n = parseInt(n, 10);
```
Should we not just validate the input instead to be either `undefined` or an integer?

addaleax replied:
@bridgear That seems like an unnecessary breaking change to me? This PR is semver-major too, but it’s less impactful, imo

addaleax commented:
@mcollina Hm, you’re right…

```
                                        confidence improvement accuracy (*)   (**)  (***)
 streams/readable-unevenread.js n=1000        ***     -4.15 %       ±1.50% ±1.99% ±2.59%
```

This does surprise me – I’ll try to look into why this is happening, it seems like an unreasonable performance difference to me.

addaleax commented:
@mcollina In the used benchmark, it is called with an integer argument, so that’s not it.
